### ISSUE TITLE
NUTS sampling performance vs. HMC

### ISSUE THREAD

**Issue by mwalmsley** [tensorflow/probability#728]
NUTS is a brilliant addition to tfp and very helpful to my research - thank you for implementing this!

I have noticed that, for the toy problem below and for my own research problems, NUTS is around 2-5x slower per sample than HMC. Is this expected?

To illustrate, below is a toy problem solved using either HMC or NUTS, and dual step size adaption (I could add this to the docs as an example if you like). 

This is using TF 1.15 with eager execution enabled.

#### Imports
```
import numpy as np
import tensorflow_probability as tfp
tfd = tfp.distributions
tf.enable_eager_execution()
```

#### Define distribution to sample (two slightly overlapping 2D Gaussians)
```
dist = tfd.MixtureSameFamily(
    mixture_distribution=tfd.Categorical(
        probs=[0.5, 0.5]),
    components_distribution=tfd.MultivariateNormalDiag(
    loc=[[.8, .5], [.2, .5]],
    scale_diag=[[.2, .4], [.2, .4]])
)
log_prob = dist.log_prob
```
 
#### Choose sampling parameters
```
n_chains = 30
initial_state = tf.constant(np.random.rand(n_chains, 2), dtype=tf.float32)
num_results = 1000
num_burnin = 100
step_size = .01  # initial
```

#### Create kernels

```
nuts_kernel = tfp.mcmc.NoUTurnSampler(
    target_log_prob_fn=log_prob,
    step_size=step_size
)

# Same num. leapfrog steps as the docstring example. 
# With step size adaption, works perfectly for this problem.
hmc_kernel = tfp.mcmc.HamiltonianMonteCarlo(
    target_log_prob_fn=log_prob,
    num_leapfrog_steps=3,
    step_size=step_size
)
```

#### Create (equivalent) step size adaption

```
hmc_adaptive_kernel = tfp.mcmc.DualAveragingStepSizeAdaptation(
    inner_kernel=hmc_kernel,
    num_adaptation_steps=int(num_burnin * 0.8)
)

nuts_adaptive_kernel = tfp.mcmc.DualAveragingStepSizeAdaptation(
    nuts_kernel,
    num_adaptation_steps=int(num_burnin * 0.8),
    step_size_setter_fn=lambda pkr, new_step_size: pkr._replace(step_size=new_step_size),
    step_size_getter_fn=lambda pkr: pkr.step_size,
    log_accept_prob_getter_fn=lambda pkr: pkr.log_accept_ratio,
)
```
(with thanks to #549)

#### Sample using each kernel

```
chain_output = tfp.mcmc.sample_chain(
    num_results=num_results,
    num_burnin_steps=num_burnin,
    current_state=initial_state,
    kernel=hmc_adaptive_kernel,
    trace_fn=None
)
```
HMC completes in 50 seconds.

```
chain_output = tfp.mcmc.sample_chain(
    num_results=num_results,
    num_burnin_steps=num_burnin,
    current_state=initial_state,
    kernel=nuts_adaptive_kernel,
    trace_fn=None
)
```
NUTS completes in 252 seconds.

**Comment by junpenglao** [comment]
Thanks for the feedback @mwalmsley, I am somewhat aware of the issue and trying to find way to fix it. So below are a couple of considerations:

- Issue of sub-optimal mass matrix, and how batched NUTS somehow amplify the problem
NUTS is very sensitive to the mass matrix of the leapfrog (which in TFP it translate to none scalar step size), and because of the batching, each step need to wait for all chains to terminate. This means if one chain is doing an extra tree doubling, you can consider that sample is going to take twice as long. You can check it by doing:

```python
chain_output_nuts, tree_size = tfp.mcmc.sample_chain(
    num_results=num_results,
    num_burnin_steps=num_burnin,
    current_state=initial_state,
    kernel=nuts_adaptive_kernel,
    trace_fn=lambda _, pkr: pkr.inner_results.leapfrogs_taken)

print(f"Number of total leapfrog taken: {tf.reduce_sum(tf.reduce_max(tree_size, axis=1))}")
print(f"Number of total leapfrog taken: {(num_results + num_burnin) * 3}")
```
which NUTS is doing around 3x more leapfrogs, so you would expect at least 3 times longer using NUTS.

- Consider number of effective samples, not just number of samples, when considering speed of a MCMC algorithm
So a sub optimal mass matrix (step size) makes NUTS slow, but not all is lost as NUTS in general return higher number of effective samples compare to other MCMC samplers

So the solutions are:
- Use XLA, which would give you immediate speed up:
```python
@tf.function(experimental_compile=True)
def do_sample_nuts():
  return tfp.mcmc.sample_chain(
    num_results=num_results,
    num_burnin_steps=num_burnin,
    current_state=initial_state,
    kernel=nuts_adaptive_kernel,
    trace_fn=lambda _, pkr: pkr.inner_results.leapfrogs_taken
)

%timeit chain_output_nuts, tree_size = do_sample_nuts()
```
- Use better step size, usually this is done by running chucks of shorter mcmc chains and estimate the covariance of the posterior samples, and use it to set the step size for the next chuck. This is an approach used by Stan and PyMC3 and you can find a demo of how to do that in [TFP here](https://colab.research.google.com/gist/junpenglao/51cd25c6372f8d2ab3490d4af8f97401/tfp_nuts_demo.ipynb#scrollTo=LDHwt0jqi211) (I am working on checking it in)

**Comment by mwalmsley** [comment]
Thanks @junpenglao for the quick and super-helpful reply!

I definitely take your point re. the leapfrog steps - I did realise this would likely be different, but didn't know how to measure the NUTS L to make the test equivalent. 

XLA is a great tip! Running CPU-only on the toy problem, with 100k steps/10k samples/128 chains, XLA takes the runtime from 2 min 11 secs (decorator only) to 8 seconds. Not bad for a one-liner!

Re. step size, thank you for the example. I'm using the new dual adaption for burn-in step size adaption, which seems to do an excellent job at getting acceptance ~ 0.75 for the main samples, but if it struggles I will follow the chuck approach you suggest :)

Closing this for now. I'll let you know how the tips play out for my research problem. Thanks again!

**Comment by junpenglao** [comment]
Happy to help!

**Comment by brianwa84** [comment]
cc @seantalts re: 3x leapfrogs

**Comment by junpenglao** [comment]
yeah long tail number of leapfrog is deadly for batched dynamic HMC (only real solution is to batch at leapfrog level).

**Comment by axch** [comment]
Another possible factor is that this example has a very cheap target density, which amplifies the relative importance of NUTS's somewhat greater per-step overhead.  If more of your overall compute is in the gradients, HMC's advantage will erode somewhat (but the longer L will remain an issue, of course).
